File: src/main/java/com/morph/engine/core/Game.java
package com.morph.engine.core;

import com.morph.engine.core.gui.ConsoleGUI;
import com.morph.engine.graphics.GLDisplay;
import com.morph.engine.graphics.GLRenderingEngine;
import com.morph.engine.input.InputMapping;
import com.morph.engine.input.KeyPress;
import com.morph.engine.input.Keyboard;
import com.morph.engine.input.Mouse;
import com.morph.engine.newgui.Container;
import com.morph.engine.newgui.Element;
import com.morph.engine.newgui.GUI;
import com.morph.engine.script.GameBehavior;
import com.morph.engine.script.ScriptSystem;
import com.morph.engine.script.debug.Console;
import com.morph.engine.util.ScriptUtils;
import org.lwjgl.glfw.GLFW;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public abstract class Game {
	protected int width, height;
	protected String title;
	protected volatile boolean isRunning = false;
	protected boolean fullscreen;
	protected float dt;

	protected IWorld world;
	protected List<GameSystem> systems = new ArrayList<>();

	// TODO: Game is a god class, consider moving below fields to different classes (behaviors to camera)

	protected HashMap<String, GameBehavior> behaviors = new HashMap<>();

	protected GLDisplay display;
	protected GLRenderingEngine renderingEngine;

	protected List<Element> guiElements = new ArrayList<>();
	private List<GUI> guis = new ArrayList<>();

	private Console console;
	private ConsoleGUI consoleGUI;

	private Camera camera = Camera.Identity.INSTANCE;

	public static final int VERSION_MAJOR = 0;
	public static final int VERSION_MINOR = 7;
	public static final int VERSION_PATCH = 0;

	public static final String VERSION_STRING = VERSION_MAJOR + "." + VERSION_MINOR + "." + VERSION_PATCH;

	private long delta;

	private InputMapping inputMapping;

	public Game(int width, int height, String title, float fps, boolean fullscreen) {
		this.width = width;
		this.height = height;
		this.title = title;
		this.dt = 1.0f / fps;
		this.fullscreen = fullscreen;
		this.console = new Console(Console.ScriptType.KOTLIN, this);
		this.consoleGUI = new ConsoleGUI(this, console, width, height);
		this.inputMapping = new InputMapping();
	}

	public void start() {
		if (isRunning)
			return;

		isRunning = true;

		init();

		long currentTime = System.nanoTime();
		double accumulator = 0.0;
		while (isRunning) {
			preUpdate();

			pollEvents();

			handleInput();
			update();

			long newTime = System.nanoTime();
			double frameTime = (newTime - currentTime) / 1000000000.0;
			currentTime = newTime;

			delta = newTime - currentTime;

			accumulator += frameTime;

			while (accumulator >= dt) {
				fixedUpdate(dt);
				accumulator -= dt;
			}

			render();
			postUpdate();
		}

		destroy();
	}

	private void preUpdate() {
		Keyboard.INSTANCE.clear();
		Mouse.INSTANCE.clear();

		preGameUpdate();

		for (GameSystem gs : systems) {
			gs.preUpdate(this);
		}

		guis.forEach(GUI::preUpdate);
		behaviors.values().forEach(GameBehavior::preUpdate);
	}

	private void postUpdate() {
		postGameUpdate();

		for (GameSystem gs : systems) {
			gs.postUpdate(this);
		}

		guis.forEach(GUI::postUpdate);
		behaviors.values().forEach(GameBehavior::postUpdate);
	}

	protected void destroy() {
		display.destroy();
		ScriptUtils.INSTANCE.stop();

		System.exit(0);
	}

	private void init() {
		ScriptUtils.INSTANCE.init(this);

		display = new GLDisplay(width, height, title);
		renderingEngine = new GLRenderingEngine(this);
		ScriptSystem scriptSystem = new ScriptSystem(this);

		addSystem(renderingEngine);
		addSystem(scriptSystem);

		display.init(this);
		display.show();

		if (fullscreen)
			display.setFullscreen(width, height);

		initGame();

		systems.forEach(GameSystem::initSystem);

		consoleGUI.init();

		// TODO: Oh my god please move this somewhere else this is evil code
//		Observable.combineLatest(
//		        events.filter(e -> e == GameAction.UPDATE),
//                Observable.concat(Observable.just(new StdMouseEvent(MouseRelease.INSTANCE, 0, 0)), Mouse.getStandardMouseEvents()),
//                Mouse.INSTANCE.getScreenMousePosition(),
//                Triple::new).subscribe(vals -> {
//                    GameAction g = vals.getFirst();
//                    StdMouseEvent m = vals.getSecond();
//                    Vector2f mousePos = vals.getThird();
//
//                    if (m.getButton() == 0 && m.getAction() == MousePress.INSTANCE)
//                        System.out.println("pressed");
//
//                    for (GUI gui : guis) {
//                        for (Element e : gui.getElements()) {
//                            switch (e.getState()) {
//                                case "IDLE":
//                                    if (mousePos != null && e.contains(mousePos)) {
//                                        if (m.getButton() == 0 && m.getAction() == MousePress.INSTANCE) {
//                                            e.setState("CLICK");
//                                        } else {
//                                            e.setState("HOVER");
//                                        }
//                                    }
//                                    break;
//                                case "HOVER":
//                                    if (mousePos != null && e.contains(mousePos)) {
//                                        if (m.getButton() == 0 && m.getAction() == MousePress.INSTANCE) {
//                                            e.setState("CLICK");
//                                        }
//                                    } else {
//                                        e.setState("IDLE");
//                                    }
//                                    break;
//                                case "CLICK":
//                                    if (m.getButton() == 0 && m.getAction() == MouseRelease.INSTANCE) {
//                                        if (mousePos != null && e.contains(mousePos))
//                                            e.setState("HOVER");
//                                        else
//                                            e.setState("IDLE");
//                                    }
//                                    break;
//                            }
//                        }
//                    }
//                }
//        );
	}

	private void update() {
		inputMapping.update();

		for (GameSystem gs : systems) {
			gs.update(this);
		}

		guis.forEach(GUI::update);
		behaviors.values().forEach(GameBehavior::update);

		Keyboard.getStandardKeyEvents().forEach(e -> {
			if (e.getAction() == KeyPress.INSTANCE && e.getKey() == GLFW.GLFW_KEY_GRAVE_ACCENT) toggleConsole();
		});
	}

	public void addSystem(GameSystem gs) {
		systems.add(gs);
	}

	public void removeSystem(GameSystem gs) {
		systems.remove(gs);
	}

	protected void pollEvents() {
		display.pollEvents();
	}

	public void stop() {
		if (!isRunning)
			return;

		isRunning = false;
	}

	public void run() {
		start();
	}

	public void addElement(Element e) {
		guiElements.add(e);
		renderingEngine.register(e);
		if (e instanceof Container) {
			((Container)e).getChildren(true).forEach(this::addElement);
		}
	}

	public void removeElement(Element e) {
		guiElements.remove(e);
		renderingEngine.unregister(e);
		if (e instanceof Container) {
			((Container)e).getChildren(true).forEach(this::removeElement);
		}
	}

	public void addElements(List<Element> e) {
		e.forEach(this::addElement);
	}

	public void removeElements(List<Element> e) {
		e.forEach(this::removeElement);
	}

	public void addGUI(GUI gui) {
		guis.add(gui);
		gui.load();
		addElements(gui.getElements());
		gui.open();
	}

	public void removeGUI(GUI gui) {
		guis.remove(gui);
		gui.unload();
		removeElements(gui.getElements());
		gui.close();
	}

	public void fixedUpdate(float dt) {
		fixedGameUpdate(dt);

		for (GameSystem gs : systems) {
			gs.fixedUpdate(this, dt);
		}

		guis.forEach(gui -> gui.fixedUpdate(dt));

		behaviors.values().forEach(b -> b.fixedUpdate(dt));
	}

	public void attachBehaviorAsync(String filename) {
		ScriptUtils.INSTANCE.getScriptBehaviorAsync(filename).subscribe(behavior -> {
			behavior.setGame(this);
			behaviors.put(filename, behavior);
			behavior.init();
			behavior.start();
		});
	}

	public void replaceBehavior(String filename, GameBehavior newBehavior) {
		System.out.println("Behavior " + filename + " has been modified.");
		newBehavior.setGame(this);
		behaviors.replace(filename, newBehavior);
		newBehavior.start();
	}

	public double getActualFPS() {
		return (1.0 / delta) * 1000000000.0;
	}

	public float getTimestep() {
	    return dt;
    }

	public abstract void initGame();
	public abstract void preGameUpdate();
	public abstract void fixedGameUpdate(float dt);
	public abstract void postGameUpdate();
	public abstract void handleInput();

	public IWorld getWorld() {
		return world;
	}

	public void setWorld(IWorld nextWorld) {
		if (world != null) {
			world.getEntities().forEach(renderingEngine::unregister);
			world.destroy();
		}

		nextWorld.init();
		nextWorld.getEntities().forEach(renderingEngine::register);
		this.world = nextWorld;
	}

	public final void render() {
		renderingEngine.render(display);
	}

	public boolean isRunning() {
		return isRunning;
	}

	public Console getConsole() {
		return console;
	}

	public void openConsole() {
		addGUI(consoleGUI);
	}

	public void closeConsole() {
		removeGUI(consoleGUI);
	}

	public void toggleConsole() {
		if (consoleGUI.isOpen()) closeConsole();
		else openConsole();
	}

	public boolean isConsoleOpen() {
		return consoleGUI.isOpen();
	}

	public void handleExitEvent() {
		System.out.println("Morph is closing...");
		stop();
	}

	public void setCamera(Camera camera) {
		this.camera = camera;
		renderingEngine.setCamera(camera);
	}

	public Camera getCamera() {
		return camera;
	}

	public int getWidth() {
		return width;
	}

	public int getHeight() {
		return height;
	}

	public InputMapping getInputMapping() {
		return inputMapping;
	}

	public void setInputMapping(InputMapping inputMapping) {
		this.inputMapping = inputMapping;
	}
}

File: src/main/kotlin/com/morph/engine/core/Game.kt
package com.morph.engine.core

import com.morph.engine.core.gui.ConsoleGUI
import com.morph.engine.graphics.GLDisplay
import com.morph.engine.graphics.GLRenderingEngine
import com.morph.engine.input.InputMapping
import com.morph.engine.input.KeyPress
import com.morph.engine.input.Keyboard
import com.morph.engine.input.Mouse
import com.morph.engine.newgui.Container
import com.morph.engine.newgui.Element
import com.morph.engine.newgui.GUI
import com.morph.engine.script.GameBehavior
import com.morph.engine.script.ScriptSystem
import com.morph.engine.script.debug.Console
import com.morph.engine.util.ScriptUtils
import org.lwjgl.glfw.GLFW

import java.util.ArrayList
import java.util.HashMap

abstract class Game(
        val width: Int = 800,
        val height: Int = 600,
        protected var title: String,
        val fps: Float,
        protected var fullscreen: Boolean
) {
    @Volatile
    private var isRunning = false
    val timestep: Float = 1.0f / fps

    var world: IWorld
        set(nextWorld) {
            if (world != null) {
                world!!.entities.forEach { renderingEngine.unregister(it) }
                world!!.destroy()
            }

            nextWorld!!.init()
            nextWorld.entities.forEach { renderingEngine.register(it) }
            field = nextWorld
        }
    protected var systems: MutableList<GameSystem> = ArrayList()

    // TODO: Game is a god class, consider moving below fields to different classes (behaviors to camera)

    protected var behaviors = HashMap<String, GameBehavior>()

    protected lateinit var display: GLDisplay
    lateinit var renderingEngine: GLRenderingEngine

    protected var guiElements: MutableList<Element> = ArrayList()
    private val guis = ArrayList<GUI>()

    val console: Console
    private val consoleGUI: ConsoleGUI

    var camera: Camera = Camera.Identity
        set(camera) {
            field = camera
            renderingEngine.setCamera(camera)
        }

    private var delta: Long = 0

    var inputMapping: InputMapping? = null

    val actualFPS: Double
        get() = 1.0 / delta * 1000000000.0

    val isConsoleOpen: Boolean
        get() = consoleGUI.isOpen

    init {
        this.width = width
        this.height = height
        this.timestep = 1.0f / fps
        this.console = Console(Console.ScriptType.KOTLIN, this)
        this.consoleGUI = ConsoleGUI(this, console, width, height)
        this.inputMapping = InputMapping()
    }

    fun start() {
        if (isRunning)
            return

        isRunning = true

        init()

        var currentTime = System.nanoTime()
        var accumulator = 0.0
        while (isRunning) {
            preUpdate()

            pollEvents()

            handleInput()
            update()

            val newTime = System.nanoTime()
            val frameTime = (newTime - currentTime) / 1000000000.0
            currentTime = newTime

            delta = newTime - currentTime

            accumulator += frameTime

            while (accumulator >= timestep) {
                fixedUpdate(timestep)
                accumulator -= timestep.toDouble()
            }

            render()
            postUpdate()
        }

        destroy()
    }

    private fun preUpdate() {
        Keyboard.clear()
        Mouse.clear()

        preGameUpdate()

        for (gs in systems) {
            gs.preUpdate(this)
        }

        guis.forEach(Consumer<GUI> { it.preUpdate() })
        behaviors.values.forEach(Consumer<GameBehavior> { it.preUpdate() })
    }

    private fun postUpdate() {
        postGameUpdate()

        for (gs in systems) {
            gs.postUpdate(this)
        }

        guis.forEach(Consumer<GUI> { it.postUpdate() })
        behaviors.values.forEach(Consumer<GameBehavior> { it.postUpdate() })
    }

    protected fun destroy() {
        display.destroy()
        ScriptUtils.stop()

        System.exit(0)
    }

    private fun init() {
        ScriptUtils.init(this)

        display = GLDisplay(width, height, title)
        renderingEngine = GLRenderingEngine(this)
        val scriptSystem = ScriptSystem(this)

        addSystem(renderingEngine)
        addSystem(scriptSystem)

        display.init(this)
        display.show()

        if (fullscreen)
            display.setFullscreen(width, height)

        initGame()

        systems.forEach(GameSystem::initSystem)

        consoleGUI.init()

        // TODO: Oh my god please move this somewhere else this is evil code
        //		Observable.combineLatest(
        //		        events.filter(e -> e == GameAction.UPDATE),
        //                Observable.concat(Observable.just(new StdMouseEvent(MouseRelease.INSTANCE, 0, 0)), Mouse.getStandardMouseEvents()),
        //                Mouse.INSTANCE.getScreenMousePosition(),
        //                Triple::new).subscribe(vals -> {
        //                    GameAction g = vals.getFirst();
        //                    StdMouseEvent m = vals.getSecond();
        //                    Vector2f mousePos = vals.getThird();
        //
        //                    if (m.getButton() == 0 && m.getAction() == MousePress.INSTANCE)
        //                        System.out.println("pressed");
        //
        //                    for (GUI gui : guis) {
        //                        for (Element e : gui.getElements()) {
        //                            switch (e.getState()) {
        //                                case "IDLE":
        //                                    if (mousePos != null && e.contains(mousePos)) {
        //                                        if (m.getButton() == 0 && m.getAction() == MousePress.INSTANCE) {
        //                                            e.setState("CLICK");
        //                                        } else {
        //                                            e.setState("HOVER");
        //                                        }
        //                                    }
        //                                    break;
        //                                case "HOVER":
        //                                    if (mousePos != null && e.contains(mousePos)) {
        //                                        if (m.getButton() == 0 && m.getAction() == MousePress.INSTANCE) {
        //                                            e.setState("CLICK");
        //                                        }
        //                                    } else {
        //                                        e.setState("IDLE");
        //                                    }
        //                                    break;
        //                                case "CLICK":
        //                                    if (m.getButton() == 0 && m.getAction() == MouseRelease.INSTANCE) {
        //                                        if (mousePos != null && e.contains(mousePos))
        //                                            e.setState("HOVER");
        //                                        else
        //                                            e.setState("IDLE");
        //                                    }
        //                                    break;
        //                            }
        //                        }
        //                    }
        //                }
        //        );
    }

    private fun update() {
        inputMapping!!.update()

        for (gs in systems) {
            gs.update(this)
        }

        guis.forEach(GUI::update)
        behaviors.values.forEach(GameBehavior::update)

        Keyboard.standardKeyEvents.forEach { e -> if (e.action === KeyPress && e.key == GLFW.GLFW_KEY_GRAVE_ACCENT) toggleConsole() }
    }

    fun addSystem(gs: GameSystem) {
        systems.add(gs)
    }

    fun removeSystem(gs: GameSystem) {
        systems.remove(gs)
    }

    protected fun pollEvents() {
        display.pollEvents()
    }

    fun stop() {
        if (!isRunning)
            return

        isRunning = false
    }

    fun run() {
        start()
    }

    fun addElement(e: Element) {
        guiElements.add(e)
        renderingEngine.register(e)
        if (e is Container) {
            e.getChildren(true).forEach{ this.addElement(it) }
        }
    }

    fun removeElement(e: Element) {
        guiElements.remove(e)
        renderingEngine.unregister(e)
        if (e is Container) {
            e.getChildren(true).forEach { this.removeElement(it) }
        }
    }

    fun addElements(e: List<Element>) {
        e.forEach{ this.addElement(it) }
    }

    fun removeElements(e: List<Element>) {
        e.forEach{ this.removeElement(it) }
    }

    fun addGUI(gui: GUI) {
        guis.add(gui)
        gui.load()
        addElements(gui.getElements())
        gui.open()
    }

    fun removeGUI(gui: GUI) {
        guis.remove(gui)
        gui.unload()
        removeElements(gui.getElements())
        gui.close()
    }

    fun fixedUpdate(dt: Float) {
        fixedGameUpdate(dt)

        for (gs in systems) {
            gs.fixedUpdate(this, dt)
        }

        guis.forEach { gui -> gui.fixedUpdate(dt) }

        behaviors.values.forEach { b -> b.fixedUpdate(dt) }
    }

    fun attachBehaviorAsync(filename: String) {
        ScriptUtils.getScriptBehaviorAsync(filename).subscribe { behavior ->
            behavior.setGame(this)
            behaviors[filename] = behavior
            behavior.init()
            behavior.start()
        }
    }

    fun replaceBehavior(filename: String, newBehavior: GameBehavior) {
        println("Behavior $filename has been modified.")
        newBehavior.setGame(this)
        behaviors.replace(filename, newBehavior)
        newBehavior.start()
    }

    abstract fun initGame()
    abstract fun preGameUpdate()
    abstract fun fixedGameUpdate(dt: Float)
    abstract fun postGameUpdate()
    abstract fun handleInput()

    fun render() {
        renderingEngine.render(display)
    }

    fun openConsole() {
        addGUI(consoleGUI)
    }

    fun closeConsole() {
        removeGUI(consoleGUI)
    }

    fun toggleConsole() {
        if (consoleGUI.isOpen)
            closeConsole()
        else
            openConsole()
    }

    fun handleExitEvent() {
        println("Morph is closing...")
        stop()
    }

    companion object {

        const val VERSION_MAJOR = 0
        const val VERSION_MINOR = 8
        const val VERSION_PATCH = 6

        const val VERSION_STRING = "$VERSION_MAJOR.$VERSION_MINOR.$VERSION_PATCH"
    }
}
